### ISSUE TITLE
qstat job state filter

### ISSUE THREAD

**Issue by dpastoor** [metrumresearchgroup/gogridengine#6]
per the qstat man page the flag `-s`

```
-s {p|r|s|z|hu|ho|hs|hd|hj|ha|h|a}[+]

          Prints only jobs in the specified state,  any  combina-
          tion  of  states is possible. -s prs corresponds to the
          regular  qstat  output  without  -s  at  all.  To  show
          recently  finished  jobs, use -s z.  To display jobs in
          user/operator/system/array-dependency hold, use the  -s
          hu/ho/hs/hd  option.  The -s ha option shows jobs which
          where submitted with the qsub -a command. qstat  -s  hj
          displays  all jobs which are not eligible for execution
          unless the job has entries in the job dependency  list.
          qstat -s h is an abbreviation for qstat -s huhohshdhjha
          and qstat -s a is an abbreviation for qstat -s psr (see
          -a, -hold_jid and -hold_jid_ad options to qsub(1)).
```

**Comment by dpastoor** [comment]
There are really two contexts we need consider here:

1) if we store all jobs (cached), then do filters at the go-level, this would be likely higher performance if lots of jobs and many different filters would be used, since couldn't cache effectively the "low" filter states where lots of jobs were returned, at least without a lot of effort

2) we _don't_ cache, but instead just return some translated request and the resulting qstat output.

We should _eventually_ include both, for now we will use strategy 1

**Comment by shairozan** [comment]
What I'm looking at for the way of doing this is as follows:

For context 1:
We already have a `GetJobs` method which just returns a giant unfiltered list jobs. I want to add a `GetJobsWithFilters` function that will take a variadic slice of functions that take a slice of jobs as input and return a slice of jobs. We'll then iterate over the functions applying them to reduce the slice down before returning it. 

For context 2:
The core `qStatFromExec` method is to be updated with a parameter (previously none). It will take a variadic map of strings to strings. If none are present, no change in functionality, but if present, we will iterate over the maps to assign the keys (switches `-u` for example) and their values (johnd,janed,johnnyd) to the arguments list before sending it to exec. If the value of the map is an empty string, we will only apply the key to the arguments list. 

The other approach for context 2 would be to create a series of types that extend slice of string, one for each option, but I feel like we'll be spending more time just managing those types than anything.

**Comment by dpastoor** [comment]
for context 1, more I think about it, probably can just take a single filter func, no need for variadic in this case

```
GetJobs(joblist, func(j JobList) {
      j.Status == 'running'
}))
```

Return all running jobs by owner Devin

```
GetJobs(joblist, func(j JobList) {
      j.Status == 'running' & j.Owner == "devinp"
}))
```

**Comment by dpastoor** [comment]
Then things like time/user/etc filters could just be provided as helper filter function factories

```
# pseudo-ish code
NewUserFilter(username string) (func(JobList) {
 return(func(j JobList) bool {
    j.Owner == username
}
}
```

**Comment by shairozan** [comment]
> for context 1, more I think about it, probably can just take a single filter func, no need for variadic in this case
> 
> ```
> GetJobs(joblist, func(j JobList) {
>       j.Status == 'running'
> }))
> ```
> 
> Return all running jobs by owner Devin
> 
> ```
> GetJobs(joblist, func(j JobList) {
>       j.Status == 'running' & j.Owner == "devinp"
> }))
> ```

It does make sense to have it as variadic and not taking any input. What we're using GetJobs for in Chokhma, for example, is just to extract all the jobs from the xml and just return the list. So it's handling the work of getting the JobInfo element from the XML, serializing it, and extracting scheduled vs non scheduled jobs. 

That's why I went with variadic. If no functions are provided, you'll get the default list. However, if you provide additional filtering functions, you can then slice it down piece by piece, allowing the function to be more flexible. 

I do like the idea of having methods to provide (in factory fashion) helpers for creating the filter functions. It keeps people from really having to even understand what the components of the Job are to filter them.

**Comment by shairozan** [comment]
> for context 1, more I think about it, probably can just take a single filter func, no need for variadic in this case
> 
> ```
> GetJobs(joblist, func(j JobList) {
>       j.Status == 'running'
> }))
> ```

I may be totally misunderstanding something, but GetJobs' role is to _produce_ a slice of jobs, so you wouldn't have job to provide to it as a parameter, unless it's just the empty struct.

It's role is to just provide the slice of jobs from the scheduled and unscheduled components of the XML, whereas other applications like Chokhmah were already doing that.

**Comment by dpastoor** [comment]
Per phone convo, we're on the same page, my pseudo-code was just that, so really we're looking at something like

```
jobs := GetJobs() # or whatever gets the array of jobs back from qstat
JobFilter(jobs, filterFunc)
# or
```

where filterFunc acts just like "any" functional programming style filter and if matching TRUE would keep that job, false discard.

**Comment by dpastoor** [comment]
implemented and accepted in #11
